handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -534,7 +534,8 @@ final int maxEncryptedPacketLength0() {
      * via other synchronized blocks.
      */
     final int calculateMaxLengthForWrap(int plaintextLength, int numComponents) {
-        return (int) min(maxWrapBufferSize, plaintextLength + (long) maxWrapOverhead * numComponents);
+        // We assume as minimum 2 bytes for the plaintextLength so we can fit an alert into it.
```

Comment:
consider adding a link to https://tools.ietf.org/html/rfc5246#section-7.2 which defines the 2 bytes size.

Reply:
would it be better to just use the minimum size for each wrap operation ... and then if there is an alert return a `BUFFER_OVERFLOW`? This would manifest in the following ways:

- if there is a non-fatal alert in the middle of writing data, we will try to write data, overflow, and have to allocate another buffer to finish the wrap operation.
- if there is an alert (fatal or not) after we finished writing data we will pick it up on the next wrap operation, or in the "outboundClose" case you added below

This suggestion is operating under the assumption that alerts are relatively rare when compared to moving data back and forth. We may end up having to allocate an extra buffer if we overflow, but again if this is assumed to be rare and will result in a few bytes saved per wrap operation. Buffer pooling may make the extra few bytes less of an issue, but not everyone can use pooling.

Which reminds me this allocation is completely internal to the `SslHandler` so we could potentially always used a pooled allocator here.

Reply:
like discussed offline I think we should just use `2` for now and then maybe investigate in a followup if there is a better way. I addressed your other comments tho

Reply:
discussed offline. My main concern with this approach is mixing the application and non-application path with a special case for ALERT data, but there maybe other non-application data that OpenSSL wants to send that is non-ALERT data which maybe larger, or we may just "get lucky" because of this extra fluff. The API for OpenSSL to communicate "I want to write non-application data, and you didn't give me enough space" seems to be call SSL_write and look for the `SSL_WANT_WRITE` error code [1]. I would prefer to rely upon this API if possible which means we may need to investigate prematurely existing the wrap loop for empty buffers [2].

[1] https://github.com/netty/netty/blob/cd7007bae2de23f00d8fa253da9e1cb9ea4e4293/handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java#L784-L797
[2] https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java#L717